Sources/TuistKit/Generator/ProjectGroups.swift

```diff
@@ -1,34 +1,59 @@
 import Basic
 import Foundation
+import TuistCore
 import xcodeproj
 
+enum ProjectGroupsError: FatalError, Equatable {
+    case missingGroup(String)
+
+    var description: String {
+        switch self {
+        case let .missingGroup(group):
+            return "Couldn't find group: \(group)"
+        }
+    }
+
+    var type: ErrorType {
+        switch self {
+        case .missingGroup:
+            return .bug
+        }
+    }
+}
+
 class ProjectGroups {
     // MARK: - Attributes
 
     let main: PBXGroup
     let products: PBXGroup
     let projectManifest: PBXGroup
-    let project: PBXGroup
     let frameworks: PBXGroup
     let playgrounds: PBXGroup?
-    let pbxproj: PBXProj
+
+    private let pbxproj: PBXProj
+    private let projectGroups: [String: PBXGroup]
+    private let knownGroups: OrderedSet<PBXGroup>
 
     // MARK: - Init
 
-    init(main: PBXGroup,
-         products: PBXGroup,
-         projectManifest: PBXGroup,
-         frameworks: PBXGroup,
-         project: PBXGroup,
-         playgrounds: PBXGroup?,
-         pbxproj: PBXProj) {
+    private init(main: PBXGroup,
+                 projectGroups: [(name: String, group: PBXGroup)],
+                 products: PBXGroup,
+                 projectManifest: PBXGroup,
+                 frameworks: PBXGroup,
+                 playgrounds: PBXGroup?,
+                 pbxproj: PBXProj) {
         self.main = main
+        self.projectGroups = Dictionary(uniqueKeysWithValues: projectGroups)
```

Comment:
Is this check necessary given that we initialize instances of this object using the static method?

Reply:
I could preprocess the `projectGroups` as a dictionary externally in the `static` method - the reasoning is here I wanted to have both a dictionary and an ordered set to maintain definition order. Alternatively, I can drop the `projectGroups` lookup and relay on `main.group(named: )` method

Reply:
I think it's fine. Let's leave it as it is.